[v25.1.x] Backport AWS Glue REST catalog support

wdberkeley opened the pull request:
Backport of #26363 

I had to resolve a trivial merge conflict in the datalake coordinator BUILD file.

Also 25.1.x uses the CMake build and a file not built by Bazel but built by CMake needed to have a function call updated.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [X] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

* none

wdberkeley committed:
cloud: parameterize credentials refresh by AWS service

Credentials for signing requests to Glue will also need to be
periodically refreshed. This change updates the credentials refresh
machinery so later it can be used with Glue.

It's a bit ad hoc, but all of this code is bit ad hoc because of how
it's trying to handle multiple cloud providers and mechanisms with a
uniform abstraction.

wdberkeley committed:
cloud_roles: parameterize signature V4 by AWS service

AWS Signature V4 incorporates the AWS service as part of the
string to sign. Until now the service has always been S3, but
to authenticate requests to Glue our sigV4 implementation must
support multiple services.

wdberkeley committed:
cloud_roles: sign request with payload hash if one exists

This changes sigV4 request signing so that it can use a payload hash
instead of a default hash that depends on the verb. This is required to
sign requests to Glue, which does not support UNSIGNED-PAYLOAD requests
(empirically).

wdberkeley committed:
cloud_roles: parameterize apply_aws_credentials by service

AWS signature V4 requires the service as part of the string to sign.
This parameterizes the applier by service.

The only service using the applier is S3. Future changes will
use the generalized credentials to authenticate requests to Glue.

wdberkeley committed:
iceberg: add support for REST sigV4 authentication

This change adds SigV4 authentication as an authentication mode for the
Iceberg REST catalog clients. There are two wrinkles:

1. The request must be signed using a string to sign that depends on
   elements of the request headers and the request payload (if the
   request has one).
2. SigV4 signing credentials are usually temporary and need to be
   refreshed.

The first issue means auth is added just before the request is sent,
when it should be otherwise complete. Bearer auth, by contrast, may be
added to the request earlier.

The second issue is thornier. This change solves it by setting up a
background refresh op like the one used by cloud storage clients. If
credentials are static, they will be set up once; otherwise a background
process will refresh them when they are almost expired.

Because there are multiple catalogs created in multiple places, this
adds a new service that manages the credentials and refreshes them in
the background, like the s3 client pool does for cloud io.

This change adds configuration properties parallel to cloud storage AWS
credentials, and allows the AWS service to be configured, just in case a
user wants to set up different credentials, or use another AWS service
that provides an Iceberg rest catalog (I don't think there are any such
services right now, though).

Note for 25.1.x backport: I fixed a trivial merge conflict in the
datalake coordinator BUILD file. Some extra test definitions were
included with the credential manager test and caused a conflict.

wdberkeley committed:
config: add iceberg_rest_base_location property

Glue does not pick a base location for you. You must specify it when
creating tables. This adds a configuration property to support Glue's
behavior. Most REST catalog implementations do not need this property to
be set.

dotnwat commented:
this looks like a manual backport (?), but i don't see any indication in the commit messages about what conflicts existed.

wdberkeley committed:
[25.1.x] Fix for backport: add missing function arg in test

Looks like there's some test code that isn't compiled by Bazel but is
compiled by CMake, and that code didn't get updated for the new
make_refresh_credentials signature.

wdberkeley commented:
> this looks like a manual backport (?), but i don't see any indication in the commit messages about what conflicts existed.

It's a manual backport. Updated the appropriate commit message and the PR description.

wdberkeley committed:
[25.1.x] CMake: add credential manager to Iceberg CMake build

We dropped CMake for Bazel after 25.1, so I needed to introduce the
credential manager to the CMake build.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/68341-->
<!--984a7661-59c0-4423-9f4a-551c0708aa29-->
<!--45a3dd82-3836-4a21-bfa5-9b00eceefc86-->
<!--title-->
## Retry command for [Build#68341](https://buildkite.com/redpanda/redpanda/builds/68341)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cluster_config_test.py::ClusterConfigTest.test_valid_settings
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--f38d4bc8-ba34-4c83-9593-83338b62d5c9-->
<!--TEST RESULTS-->
<!--30d42c4a-cfc7-40f0-8c67-ab32ae487aef-->
<!--TEST RESULTS-->
<!--39840df6-7725-4ad6-93d5-9e8cf6b41f0a-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68341">build#68341</a></summary>


| test_class                    | test_method                   | test_arguments                                                     | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:------------------------------|:------------------------------|:-------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| gtest_cloud_storage_rpfixture | gtest_cloud_storage_rpfixture |                                                                    | unit        | https://buildkite.com/redpanda/redpanda/builds/68341#0197cc78-e0d7-4272-a744-24333a98057b | FLAKY         | 1/2      |                                                                                                                                                           |
| ClusterConfigTest             | test_valid_settings           |                                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68341#0197ccd0-5f5f-4c4b-8b19-ac43b50e7452 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                    |
| ClusterConfigTest             | test_valid_settings           |                                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68341#0197ccd2-932c-4b3d-a11a-976ac7773bc9 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                    |
| TestTieredStoragePause        | test_safe_pause_resume        | {"allow_gaps_cluster_level": true, "allow_gaps_topic_level": null} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68341#0197ccd2-932b-47d2-b744-f552673c3748 | FLAKY         | 20/21    | upstream reliability is '95.0'. current run reliability is '95.23809523809523'. drift is -0.2381 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68363">build#68363</a></summary>


| test_class                    | test_method                      | test_arguments                                                                                                                                                   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                      |
|:------------------------------|:---------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------|
| gtest_cloud_storage_rpfixture | gtest_cloud_storage_rpfixture    |                                                                                                                                                                  | unit        | https://buildkite.com/redpanda/redpanda/builds/68363#0197cd3c-ec6a-4e23-a619-94b110295d1f | FLAKY         | 1/2      |                                                                                                                                                             |
| kafka_server_rpfixture        | kafka_server_rpfixture           |                                                                                                                                                                  | unit        | https://buildkite.com/redpanda/redpanda/builds/68363#0197cd3c-ec6a-4e23-a619-94b110295d1f | FLAKY         | 1/2      |                                                                                                                                                             |
| ClusterConfigTest             | test_valid_settings              |                                                                                                                                                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68363#0197cd7d-bca8-4821-b46f-77baef57f9b2 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      |
| ClusterConfigTest             | test_valid_settings              |                                                                                                                                                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68363#0197cd93-e93a-4531-bb88-d900bcf52be6 | FAIL          | 0/21     | The test has failed across all retries                                                                                                                      |
| RandomNodeOperationsTest      | test_node_operations             | {"cloud_storage_type": 2, "enable_failures": true, "mixed_versions": false, "with_chunked_compaction": true, "with_iceberg": false, "with_tiered_storage": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68363#0197cd7d-bca8-4d6e-8f59-b04be531acf7 | FLAKY         | 18/21    | upstream reliability is '100.0'. current run reliability is '85.71428571428571'. drift is 14.28571 and the allowed drift is set to 50. The test should PASS |
| TieredStorageTest             | test_tiered_storage              | {"cloud_storage_type_and_url_style": [1, "path"], "test_case": {"name": "(TS_Read == True, TS_TxRangeMaterialized == True, SpilloverManifestUploaded == True)"}} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68363#0197cd93-e93a-4531-bb88-d900bcf52be6 | FAIL          | 0/1      | The test has failed across all retries                                                                                                                      |
| CompactionWithRecoveryTest    | test_tx_compaction_with_recovery |                                                                                                                                                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68363#0197cd7d-bca8-4821-b46f-77baef57f9b2 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS   |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/68444">build#68444</a></summary>


| test_class                      | test_method                     | test_arguments                                                                                                                                                    | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                    |
|:--------------------------------|:--------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------|
| distributed_kv_stm_tests_rpunit | distributed_kv_stm_tests_rpunit |                                                                                                                                                                   | unit        | https://buildkite.com/redpanda/redpanda/builds/68444#0197d194-4dd4-4806-910d-ab9faf70587b | FLAKY         | 1/2      |                                                                                                                                                           |
| RandomNodeOperationsTest        | test_node_operations            | {"cloud_storage_type": 1, "enable_failures": true, "mixed_versions": false, "with_chunked_compaction": false, "with_iceberg": false, "with_tiered_storage": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/68444#0197d1d5-6097-4776-b149-fc7f788be0bc | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

wdberkeley committed:
[25.1.x] config: remove glue auth validation check

This was needed upstream as well.

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/68363-->
<!--efcc32b7-cee2-4b1a-8719-ed079d36054c-->
<!--f6792d8e-5d29-42d4-a3a7-bc0f76366d66-->
<!--title-->
## Retry command for [Build#68363](https://buildkite.com/redpanda/redpanda/builds/68363)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cluster_config_test.py::ClusterConfigTest.test_valid_settings
tests/rptest/tests/tiered_storage_model_test.py::TieredStorageTest.test_tiered_storage@{"cloud_storage_type_and_url_style":[1,"path"],"test_case":{"name":"(TS_Read == True, TS_TxRangeMaterialized == True, SpilloverManifestUploaded == True)"}}
```
<!--/body-->

nvartolomei commented on src/v/config/configuration.cc:
```diff
@@ -3957,20 +3966,89 @@ configuration::configuration()
+      "azure_vm_instance_metadata, azure_aks_oidc_federation.",
+      {.needs_restart = needs_restart::yes,
+       .visibility = visibility::user,
+       .gets_restored = gets_restored::no},
```
didn't we change this in the end to gets_restored::yes?

wdberkeley replied:
Yeah, in a follow-up to the original PR. There's a short line of backports to go on top of this one. We could combine them. Not sure what the usual strategy is.

wdberkeley committed:
[25.11.x] Add example iceberg credentials source

cluster_config_test's test_valid_settings handled enum options
differently in 25.1. This adds an example config option so the test can
use it.
